File: packages/core/src/engine/verbs/fetch.ts
/*!
 * Copyright (c) Microsoft. All rights reserved.
 * Licensed under the MIT license. See LICENSE file in the project.
 */

import { loadCSV } from 'arquero'
import type ColumnTable from 'arquero/dist/types/table/column-table'
import type { FetchArgs, Step, TableStore } from '../../index.js'

/**
 * Executes an arquero loadCSV
 * @param step
 * @param store
 * @returns
 */
export async function fetch(
	step: Step,
	_store: TableStore,
): Promise<ColumnTable> {
	const { args } = step
	const { url, delimiter } = args as FetchArgs

	const tableFromCSV: Promise<ColumnTable> = loadCSV(url, {
		delimiter: delimiter,
	})
	return tableFromCSV
}

File: python/data-wrangling-components/data_wrangling_components/engine/verbs/fetch.py
#
# Copyright (c) Microsoft. All rights reserved.
# Licensed under the MIT license. See LICENSE file in the project.
#

from urllib.parse import urlparse

import pandas as pd

from data_wrangling_components.table_store import TableContainer


__reader_mapping = {
    "csv": lambda **kwargs: pd.read_csv(
        kwargs["url"], sep=kwargs["delimiter"], na_values=["undefined"]
    ),
    "json": lambda **kwargs: pd.read_json(kwargs["url"]),
}


def fetch(url: str, delimiter: str = ",", **kwargs):
    """Fetch a table from a URL.

    :param step:
        Parameters to execute the operation.
        See :py:class:`~data_wrangling_components.engine.verbs.fetch.FetchArgs`.
    :type step: Step
    :param store:
        Table store that contains the inputs to be used in the execution.
    :type store: TableStore

    :return: new table with the result of the operation
    """
    file_type = urlparse(url).path.split(".")[-1]
    output = __reader_mapping[file_type](**{"url": url, "delimiter": delimiter})
    return TableContainer(table=output)
